docs/source/json-api/production-setup.rst

```diff
@@ -0,0 +1,300 @@
+
+To protect data in transit and over untrusted networks, the *HTTP-JSON API* server provides
+TLS support, to enable TLS you need to specify the private key for your server and the
+certificate chain via ``daml json-api --pem server.pem --crt server.crt``. You can also
+set a custom root CA certificate used to validate client certificates via ``--cacrt ca.crt``
+
+
+Architecture
+************
+
+Components
+----------
+
+A production setup of the *HTTP-JSON API* will involve the following components:
+
+- the *HTTP-JSON API* server
+- the query store backend database server
+- the ledger
+
+*HTTP-JSON API* server exposes an API to interact with the Ledger and it uses JDBC to interact
+with its underlying query store for caching and serving data efficiently.
+
+The *HTTP-JSON API* server releases are regularly tested with OpenJDK 8 on a 64-bit x86 architecture,
+with Ubuntu 20.04, macOS 11.5.2 and Windows Server 2016.
+
+In production, we recommend running on a 64-bit x86 architecture in a Linux
+environment. This environment should have a Java SE Runtime Environment such
+as OpenJDK JRE and must be compatible with OpenJDK version 1.8.0_202 or later.
+We recommend using PostgreSQL server as query-store, most of our tests have
+been done with servers running version > 10.
+
+
+Scaling and Redundancy
+**********************
+
+.. note:: This section of the document only talks about scaling and redundancy setup for the *HTTP-JSON API* server. In all of the recommendations suggested below we assume that the JSON API always interacts with a single participant on the ledger.
+
+We advise that the *HTTP-JSON API* server and *query store* components to have dedicated
+computation and memory resources available to them. This can be achieved via
+containerization or setting them up on independent physical servers. Ensure that the two
+components are **physically co-located** to reduce network latency for
+communication. The scaling and availability aspects heavily rely on the interactions between
+the core components listed above.
+
+With respect to scaling we recommend to follow the general advice in trying to
+understand the bottlenecks and see if adding additional processing power/memory
+i.e vertical scaling is beneficial.
+
+In general for horizontal scaling purposes , we recommend to treat the
+*HTTP-JSON API* and the *query store database server* as a single unit for scaling.
+While a setup with multiple *HTTP-JSON API* services running against a single backend
+database server is feasible, it maybe a futile exercise if the database server
+itself is the bottleneck.
+
+We can run a redundant setup for the *HTTP-JSON API* by using a reverse proxy server with some
+acceptable routing mechanism and then and having multiple *HTTP-JSON API* servers sit behind it
+each with their own backend database servers and dedicated computation and memory resources.
```

Comment:
This is where I would expect some explanation of exactly how the query store is constructed and from that what an "acceptable routing mechanism" is.

Reply:
@garyverhaegen-da not sure what you mean when you say "how the query store is constructed" ? w.r.t the routing mechanism are you hinting towards suggesting sticky sessions based on parties querying the contract dataset ? I think those decisions can be left to the operators of the deployment.

Reply:
> I think those decisions can be left to the operators of the deployment.

Decisions, sure, but we should make sure they have the information needed. If we don't tell them why they might want to, they can't really make an informed decision.

I'm not familiar enough with the JSON API to be confident any of the following is true, but I would expect something along these lines:

> The Query Store is built by saving the state of the ACS up to the current offset, which allows the JSON API to only request the delta on subsequent queries, making it much faster than having to request the entire ACS every time. Because of the Daml privacy model (JSON API requests are made using the user-provided token), the data stored in a given Query Store will be specific to the set of parties that have made queries through that specific Query Store instance (for a given template). Therefore, if you do run with separate Query Stores, it may be useful to route queries based on requesting party (and possibly queried template), which would minimize the amount of data in each Query Store as well as the overall redundancy of said data.

Which reminds me, it's also not very clear from this document why the operator needs to be careful with data privacy, which this explanation would also somewhat cover.

Reply:
I have split this into 2 and added parts about how the query store is built to the top `Query Store` section and the other bits here